llama-index-integrations/storage/chat_store/llama-index-storage-chat-store-dynamodb/llama_index/storage/chat_store/dynamodb/base.py

```diff
@@ -0,0 +1,282 @@
+        """
+        response = self._table.get_item(Key={self.primary_key: key})
+
+        if response and "Item" in response:
+            message_history = response["Item"]["History"]
+        else:
+            message_history = []
+
+        return [_dict_to_message(message) for message in message_history]
+
+    def add_message(self, key: str, message: ChatMessage) -> None:
+        """Add a message to the end of the chat history for the given key.
+        Creates a new row if the key does not exist.
+
+        Args:
+            key (str): The key specifying a row.
+            message (ChatMessage): The message to add to the chat history.
+
+        Returns:
+            None
+        """
+        current_messages = _messages_to_dict(self.get_messages(key))
+        current_messages.append(_message_to_dict(message))
+
+        self._table.put_item(Item={self.primary_key: key, "History": current_messages})
+
+    def delete_messages(self, key: str) -> Optional[List[ChatMessage]]:
+        """Deletes the entire chat history for the given key (i.e. the row).
+
+        Args:
+            key (str): The key specifying a row.
+
+        Returns:
+            Optional[List[ChatMessage]]: The messages that were deleted. None if the
+                deletion failed.
+        """
+        messages_to_delete = self.get_messages(key)
+        self._table.delete_item(Key={self.primary_key: key})
+        return messages_to_delete
+
+    def delete_message(self, key: str, idx: int) -> Optional[ChatMessage]:
+        """Deletes the message at the given index for the given key.
+
+        Args:
+            key (str): The key specifying a row.
+            idx (int): The index of the message to delete.
+
+        Returns:
+            Optional[ChatMessage]: The message that was deleted. None if the index
+                did not exist.
+        """
+        current_messages = self.get_messages(key)
+        try:
+            message_to_delete = current_messages[idx]
+            del current_messages[idx]
+            self.set_messages(key, current_messages)
+            return message_to_delete
+        except IndexError:
+            logger.error(
+                IndexError(f"No message exists at index, {idx}, for session {key}")
+            )
+            return None
+
+    def delete_last_message(self, key: str) -> Optional[ChatMessage]:
+        """Deletes the last message in the chat history for the given key.
+
+        Args:
+            key (str): The key specifying a row.
+
+        Returns:
+            Optional[ChatMessage]: The message that was deleted. None if the chat history
+                was empty.
+        """
+        return self.delete_message(key, -1)
+
+    def get_keys(self) -> List[str]:
+        """Retrieve all keys in the table.
+
+        Returns:
+            List[str]: The keys in the table.
+        """
+        response = self._table.scan(ProjectionExpression="SessionId")
```

Comment:
should we be using `self.primary_key` instead of `SessionID` here?

Reply:
Good catch. Corrected.